mermithid/processors/misc/BinnedDataFitter.py

```diff
@@ -0,0 +1,169 @@
+        self.parameter_errors = [max([0.1, 0.1*p]) for p in self.initial_values]
+
+        return True
+
+    def InternalRun(self):
+        logger.info('namedata: {}'.format(self.namedata))
+
+        self.hist, _ = np.histogram(self.data[self.namedata], self.bins)
+
+        result_array, error_array = self.fit()
+
+        # save results
+        self.results = {}
+        self.results['param_values'] = result_array
+        self.results['param_errors'] = error_array
+        for i, k in enumerate(self.parameter_names):
+            self.results[k] = {'value': result_array[i], 'error': error_array[i]}
+
+        return True
+
+
+
+    def PDF(self, x, A, mu, sigma):
+        """
+        Overwrite by whatever PDF
+        """
+        f = A*(1/(sigma*np.sqrt(2*np.pi)))*np.exp(-(((x-mu)/sigma)**2.)/2.)
+        return f
+
+
+
+    def fit(self):
+        # Now minimize neg log likelihood using iMinuit
+        if self.print_level == 1:
+            logger.info('This is the plan:')
+            logger.info('Fitting data consisting of {} elements'.format(np.sum(self.hist)))
+            logger.info('Fit parameters: {}'.format(self.parameter_names))
+            logger.info('Initial values: {}'.format(self.initial_values))
+            logger.info('Initial error: {}'.format(self.parameter_errors))
+            logger.info('Fixed in fit: {}'.format(self.fixes))
+            logger.info('Constrained parameters: {}'.format([self.parameter_names[i] for i in self.constrained_parameters]))
+
+        m_binned = Minuit.from_array_func(self.negPoissonLogLikelihood,
+                                      self.initial_values,
+                                      error=self.parameter_errors,
+                                      errordef = 0.5, limit = self.limits,
+                                      name=self.parameter_names,
+                                      fix=self.fixes,
+                                      print_level=self.print_level,
+                                      throw_nan=True
+                                      )
+
+
+        # minimze
+        m_binned.migrad(resume=False)
+        self.param_states = m_binned.get_param_states()
+        self.m_binned = m_binned
+
+        # results
+        result_array = m_binned.np_values()
+        error_array = m_binned.np_errors()
+
+        if self.print_level == 1:
+            logger.info('Fit results: {}'.format(result_array))
+            logger.info('Errors: {}'.format(error_array))
+        return result_array, error_array
+
+
+    def PoissonLogLikelihood(self, params):
+
+        # binned data
+        hist = self.hist
+
+        # expectation
+        pdf_return = self.PDF(self.bin_centers, *params)
+        if np.shape(pdf_return)[0] == 2:
+            expectation, expectation_error = pdf_return
+        else:
+            expectation = pdf_return
+
+        # exclude bins where expectation is <= zero or nan
+        index = np.where(expectation>0)
```

Comment:
Good check; but if there are expected predicted number of events in a bin which is negative, you are in a lot of trouble...
There should be a huge error message if you have negative expectations.

Reply:
added error message but did not raise exception. is that an ok solution?

Reply:
Sure, the fact that this error message will be printed a lot of times (as many as the number of calls to this function) sounds like a bug enough warning for a not-super-careful user! ;)